File: compilers/python/tests/test_constructor_args_validation.py
"""constructorArgs shape validation.

Mirrors packages/runar-compiler/src/__tests__/constructor-args-validation.test.ts.

``compile_from_source(constructor_args=...)`` must reject inputs that would
silently bake nothing and emit placeholder scripts that fail opaquely at
runtime:

 (a) positional arrays (natural guess, but keys match no property names);
 (b) keys that don't match any contract property (typos);
 (c) referenced readonly properties left unbaked after applying the args.

Python is dynamically typed, so unlike Go/Rust the positional-array reject (a)
is a real, reachable case and must be checked. snake_case source identifiers
map to camelCase property names in the ANF, so keys are camelCase.
"""

from __future__ import annotations

import textwrap

import pytest

from runar_compiler.compiler import CompilationError, compile_from_source


HASH_LOCK_SOURCE = textwrap.dedent("""\
    from runar import SmartContract, ByteString, Sha256, public, assert_, sha256

    class HashLock(SmartContract):
        hash_value: Sha256

        def __init__(self, hash_value: Sha256):
            super().__init__(hash_value)
            self.hash_value = hash_value

        @public
        def unlock(self, preimage: ByteString):
            assert_(sha256(preimage) == self.hash_value)
""")

TWO_PROP_SOURCE = textwrap.dedent("""\
    from runar import SmartContract, Bigint, public, assert_

    class TwoProp(SmartContract):
        target: Bigint
        unused: Bigint

        def __init__(self, target: Bigint, unused: Bigint):
            super().__init__(target, unused)
            self.target = target
            self.unused = unused

        @public
        def check(self, x: Bigint):
            assert_(x == self.target)
""")

HASH = "aa" * 32


def _compile(tmp_path, source: str, name: str, constructor_args=None):
    path = tmp_path / name
    path.write_text(source, encoding="utf-8")
    return compile_from_source(str(path), constructor_args=constructor_args)


class TestConstructorArgsValidation:
    def test_rejects_positional_array(self, tmp_path):
        # Positional array -- the shape RunarContract/TestContract take, but NOT
        # what the compiler takes. Previously baked nothing silently.
        with pytest.raises(CompilationError, match="positional array"):
            _compile(tmp_path, HASH_LOCK_SOURCE, "HashLock.runar.py",
                     constructor_args=[HASH])

    def test_rejects_keys_matching_no_property(self, tmp_path):
        with pytest.raises(CompilationError) as exc:
            _compile(tmp_path, HASH_LOCK_SOURCE, "HashLock.runar.py",
                     constructor_args={"hashVal": HASH})  # typo: hashValue
        msg = str(exc.value)
        assert "'hashVal'" in msg
        assert "hashValue" in msg

    def test_rejects_referenced_readonly_left_unbaked(self, tmp_path):
        # 'target' is referenced by check() but not provided.
        with pytest.raises(CompilationError, match=r"'target'.*placeholder"):
            _compile(tmp_path, TWO_PROP_SOURCE, "TwoProp.runar.py",
                     constructor_args={"unused": 1})

    def test_accepts_unreferenced_readonly_left_unbaked(self, tmp_path):
        # 'unused' is never referenced by a method -- DCE eliminates it, so
        # leaving it unbaked is fine.
        artifact = _compile(tmp_path, TWO_PROP_SOURCE, "TwoProp.runar.py",
                            constructor_args={"target": 42})
        assert len(artifact.script) > 0

    def test_accepts_complete_named_record(self, tmp_path):
        artifact = _compile(tmp_path, HASH_LOCK_SOURCE, "HashLock.runar.py",
                            constructor_args={"hashValue": HASH})
        assert HASH in artifact.script
        # Fully baked -- no placeholder constructor slots.
        assert len(artifact.constructor_slots) == 0

    def test_still_compiles_placeholder_when_no_args(self, tmp_path):
        artifact = _compile(tmp_path, HASH_LOCK_SOURCE, "HashLock.runar.py")
        assert len(artifact.constructor_slots) >= 1

File: compilers/ruby/test/test_constructor_args_validation.rb
# frozen_string_literal: true

require_relative "test_helper"

require "runar_compiler/frontend/ast_nodes"
require "runar_compiler/frontend/diagnostic"
require "runar_compiler/frontend/validator"
require "runar_compiler/frontend/typecheck"
require "runar_compiler/frontend/anf_lower"
require "runar_compiler/frontend/parser_ts"

# constructor_args shape validation -- `_apply_constructor_args` must reject
# inputs that would silently bake nothing and emit placeholder scripts that
# fail opaquely at runtime:
#
#  (a) positional arrays (natural guess, but keys match no property names)
#  (b) keys that don't match any contract property (typos)
#  (c) referenced readonly properties left unbaked after applying the args
class TestConstructorArgsValidation < Minitest::Test
  HASH_LOCK_SOURCE = <<~TS
    class HashLock extends SmartContract {
      readonly hashValue: Sha256;

      constructor(hashValue: Sha256) {
        super(hashValue);
        this.hashValue = hashValue;
      }

      public unlock(preimage: ByteString) {
        assert(sha256(preimage) === this.hashValue);
      }
    }
  TS

  TWO_PROP_SOURCE = <<~TS
    class TwoProp extends SmartContract {
      readonly target: bigint;
      readonly unused: bigint;

      constructor(target: bigint, unused: bigint) {
        super(target, unused);
        this.target = target;
        this.unused = unused;
      }

      public check(x: bigint) {
        assert(x === this.target);
      }
    }
  TS

  HASH = "aa" * 32

  def build_program(source, file_name = "Test.runar.ts")
    parse_result = RunarCompiler.send(:_parse_source, source, file_name)
    assert_empty parse_result.errors.map(&:format_message), "unexpected parse errors"
    refute_nil parse_result.contract
    RunarCompiler::Frontend.lower_to_anf(parse_result.contract)
  end

  # Apply args (raises on invalid) then compile.
  def compile_with_args(source, args, file_name = "Test.runar.ts")
    program = build_program(source, file_name)
    RunarCompiler.send(:_apply_constructor_args, program, args)
    RunarCompiler.compile_from_program(program, disable_constant_folding: true)
  end

  def apply_args(source, args, file_name = "Test.runar.ts")
    program = build_program(source, file_name)
    RunarCompiler.send(:_apply_constructor_args, program, args)
  end

  def test_rejects_a_positional_array
    # Positional array -- the shape RunarContract takes, but NOT what the
    # compiler baking path takes. Previously baked nothing silently.
    err = assert_raises(RunarCompiler::CompilationError) do
      apply_args(HASH_LOCK_SOURCE, [HASH], "HashLock.runar.ts")
    end
    assert_match(/positional array/i, err.message)
  end

  def test_rejects_keys_that_match_no_contract_property
    err = assert_raises(RunarCompiler::CompilationError) do
      apply_args(HASH_LOCK_SOURCE, { "hashVal" => HASH }, "HashLock.runar.ts") # typo: hashValue
    end
    assert_includes err.message, "'hashVal'"
    assert_includes err.message, "hashValue"
  end

  def test_rejects_when_a_referenced_readonly_property_remains_unbaked
    # 'target' is referenced by check() but not provided.
    err = assert_raises(RunarCompiler::CompilationError) do
      apply_args(TWO_PROP_SOURCE, { "unused" => 1 }, "TwoProp.runar.ts")
    end
    assert_match(/'target'/, err.message)
    assert_match(/placeholder/, err.message)
  end

  def test_accepts_an_unreferenced_readonly_property_left_unbaked
    # 'unused' is never referenced by a method -- DCE eliminates it, so
    # leaving it unbaked is fine.
    artifact = compile_with_args(TWO_PROP_SOURCE, { "target" => 42 }, "TwoProp.runar.ts")
    refute_nil artifact.script
    refute_empty artifact.script
  end

  def test_accepts_a_complete_named_record
    artifact = compile_with_args(HASH_LOCK_SOURCE, { "hashValue" => HASH }, "HashLock.runar.ts")
    assert_includes artifact.script, HASH
    slots = artifact.constructor_slots
    assert(slots.nil? || slots.empty?, "expected no constructor slots for a fully baked record")
  end

  def test_still_compiles_placeholder_artifacts_when_no_args_given
    program = build_program(HASH_LOCK_SOURCE, "HashLock.runar.ts")
    RunarCompiler.send(:_apply_constructor_args, program, nil)
    artifact = RunarCompiler.compile_from_program(program, disable_constant_folding: true)
    assert_operator artifact.constructor_slots.length, :>=, 1
  end

  def test_empty_record_is_the_unchecked_placeholder_path
    # An empty record bakes nothing and must not trip the referenced-readonly
    # check (byte-identical to the no-args placeholder path).
    program = build_program(HASH_LOCK_SOURCE, "HashLock.runar.ts")
    RunarCompiler.send(:_apply_constructor_args, program, {})
    artifact = RunarCompiler.compile_from_program(program, disable_constant_folding: true)
    assert_operator artifact.constructor_slots.length, :>=, 1
  end
end
